src/bin/decode.rs

```diff
@@ -45,40 +49,44 @@ fn main() {
         ])
     .get_matches();
 
-    let quiet = matches.is_present("quiet");
-    let source_path = matches.value_of("INPUT")
-        .expect("Expected input file");
-    let dest_path = matches.value_of("OUTPUT")
-        .expect("Expected output file");
+    let source_path = matches.value_of("INPUT");
+    let dest_path = matches.value_of("OUTPUT");
+    let quiet = matches.is_present("quiet") || dest_path.is_none();
 
     // Setup.
-    let printer = Shift::new();
+    let options = Options {
+        print_json: matches.is_present("print-json"),
+        quiet,
+        dest_path,
+    };
 
     progress!(quiet, "Reading.");
-    let file = File::open(source_path)
-        .expect("Could not open source");
-    let stream = BufReader::new(file);
-
-    progress!(quiet, "Attempting to decode as multipart.");
-    let tree : binjs::specialized::es6::ast::Script = if let Ok(reader) = binjs::io::multipart::TreeTokenReader::new(stream) {
-        let mut deserializer = binjs::specialized::es6::io::Deserializer::new(reader);
-        deserializer.deserialize(&mut binjs::specialized::es6::ast::IOPath::new())
-            .expect("Could not decode")
-    } else {
-        progress!(quiet, "... falling back to simple format.");
-
-        let file = File::open(source_path)
-            .expect("Could not open source");
-        let stream = BufReader::new(file);
-
-        let reader = binjs::io::simple::TreeTokenReader::new(stream);
-        let mut deserializer = binjs::specialized::es6::io::Deserializer::new(reader);
-        deserializer.deserialize(&mut binjs::specialized::es6::ast::IOPath::new())
-            .expect("Could not decode")
+    let tree : binjs::specialized::es6::ast::Script = match source_path {
+        Some(path) => {
+            parse_tree(&|| {
+                let file = File::open(path)
+                    .expect("Could not open source");
+                BufReader::new(file)
+            }, &options)
+        }
+        None => {
+            let mut buffer = Vec::new();
+            stdin().read_to_end(&mut buffer)
```

Comment:
Looks like you need to go through many hoops because we need to implement `Seek` for our input stream. I seem to remember that `Seek` is only ever used for error messages, so could you try and see if there is a way to get rid of that dependency?

This would simplify your code, and it's not a big deal if we lose a few offsets in error messages at this stage.

If you want to land this before getting rid of `Seek`, that would also work, but in that case, could you file an issue?

Reply:
yes, I'll file a bug, to keep the diff small and scoped.